lib/executor/base_executor.go

```diff
@@ -38,53 +40,74 @@ import (
 type BaseExecutor struct {
 	config         lib.ExecutorConfig
 	executionState *lib.ExecutionState
+	VUIDLocal      *uint64 // counter for assigning executor-specific VU IDs
+	iterSegIndexMx *sync.Mutex
+	iterSegIndex   *lib.SegmentedIndex
 	logger         *logrus.Entry
 	progress       *pb.ProgressBar
 }
 
 // NewBaseExecutor returns an initialized BaseExecutor
 func NewBaseExecutor(config lib.ExecutorConfig, es *lib.ExecutionState, logger *logrus.Entry) *BaseExecutor {
+	segIdx := lib.NewSegmentedIndex(es.ExecutionTuple)
 	return &BaseExecutor{
 		config:         config,
 		executionState: es,
+		VUIDLocal:      new(uint64),
 		logger:         logger,
+		iterSegIndexMx: new(sync.Mutex),
+		iterSegIndex:   segIdx,
 		progress: pb.New(
 			pb.WithLeft(config.GetName),
 			pb.WithLogger(logger),
 		),
 	}
 }
 
+// nextIterationCounters next scaled(local) and unscaled(global) iteration counters
+func (bs *BaseExecutor) nextIterationCounters() (uint64, uint64) {
+	bs.iterSegIndexMx.Lock()
+	defer bs.iterSegIndexMx.Unlock()
+	scaled, unscaled := bs.iterSegIndex.Next()
+	return uint64(scaled - 1), uint64(unscaled - 1)
+}
+
 // Init doesn't do anything for most executors, since initialization of all
 // planned VUs is handled by the executor.
 func (bs *BaseExecutor) Init(_ context.Context) error {
 	return nil
 }
 
 // GetConfig returns the configuration with which this executor was launched.
-func (bs BaseExecutor) GetConfig() lib.ExecutorConfig {
+func (bs *BaseExecutor) GetConfig() lib.ExecutorConfig {
 	return bs.config
 }
 
+// getNextLocalVUID increments and returns the next VU ID that's specific for
+// this executor (i.e. not global like __VU).
+func (bs *BaseExecutor) getNextLocalVUID() uint64 {
+	return atomic.AddUint64(bs.VUIDLocal, 1)
```

Comment:
Shouldn't this also be segmented? Right now with it being per-instance, I am not sure how useful it will be?

Reply:
:-/ We brought up actually removing this counter in [this thread](https://github.com/k6io/k6/pull/1863#discussion_r651867771). I'm not sure how helpful it is already, but even less so if we add another global variant of it. I think we already have a lot of possibly confusing values and adding anything new without knowing if users will actually need it only forces us to support it on the off chance someone is depending on it. The module is already large enough, so I'd rather leave this as is or remove it and add it later if needed.

Reply:
Ah, I replied in that previous thread only to your latest message, without reading the start... :sweat_smile: :+1: for removing this counter and potentially adding it back in the future. I could argue why a segmented per-scenario VU index in would be useful, but we can certainly add it later.

Reply:
Hmm actually, a segmented index here would be most useful for things like `constant-vus` and `ramping-vus` and maybe `per-vu-iterations`, where the VU number is the leading thing, not the iteration number, so that's what people would likely want to segment on :man_shrugging: 

So, yeah, we can remove it for now, so we can ship the PR more quickly, but I'm almost certain we're going to have to put a segmented per-scenario VU index back fairly soon :sweat_smile: Especially since we still don't have a per-scenario VU init function (https://github.com/k6io/k6/issues/785) where something like this could be generated easily by users.

Reply:
OK, I removed it in 670b7126 :disappointed: FYI @MStoykov